src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp

```diff
@@ -0,0 +1,982 @@
+
+#pragma once
+
+#include <boost/functional/hash.hpp>  // IWYU pragma: keep
+#include <cmath>
+#include <cstddef>
+#include <limits>
+#include <ostream>
+#include <pup.h>  // IWYU pragma: keep
+#include <string>
+#include <type_traits>
+#include <unordered_map>
+#include <utility>  // for pair
+#include <vector>
+
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "DataStructures/DataVector.hpp"  // IWYU pragma: keep
+#include "DataStructures/Index.hpp"
+#include "DataStructures/ModalVector.hpp"           // IWYU pragma: keep
+#include "DataStructures/Tensor/Metafunctions.hpp"  // IWYU pragma: keep
+#include "DataStructures/Tensor/Tensor.hpp"         // IWYU pragma: keep
+#include "DataStructures/Variables.hpp"             // IWYU pragma: keep
+#include "DataStructures/VariablesHelpers.hpp"
+#include "Domain/Direction.hpp"
+#include "Domain/Element.hpp"    // IWYU pragma: keep
+#include "Domain/ElementId.hpp"  // IWYU pragma: keep
+#include "Domain/Mesh.hpp"
+#include "Domain/OrientationMap.hpp"  // IWYU pragma: keep
+#include "Domain/Tags.hpp"            // IWYU pragma: keep
+#include "ErrorHandling/Assert.hpp"
+#include "ErrorHandling/Error.hpp"
+#include "NumericalAlgorithms/LinearOperators/CoefficientTransforms.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "Options/Options.hpp"
+#include "Utilities/Algorithm.hpp"
+#include "Utilities/ForceInline.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/Math.hpp"
+#include "Utilities/TMPL.hpp"
+
+// IWYU pragma: no_include <algorithm>
+// IWYU pragma: no_forward_declare Variables
+
+/// \cond
+namespace SlopeLimiters {
+template <size_t VolumeDim, typename TagsToLimit>
+class Hierarchical;
+}  // namespace SlopeLimiters
+/// \endcond
+
+namespace SlopeLimiters {
+enum class HierarchicalType { Krivodonova };
+
+std::ostream& operator<<(std::ostream& os, HierarchicalType t) noexcept;
+
+/// Given the `NodalTag` swap the `DataVector` with a `ModalVector`.
+template <typename NodalTag>
+struct ModalTag : db::SimpleTag {
+  using type =
+      TensorMetafunctions::swap_type<ModalVector, typename NodalTag::type>;
+  static std::string name() noexcept {
+    return "coefficients(" + NodalTag::name() + ")";
+  }
+};
+
+/*!
+ * \ingroup SlopeLimitersGroup
+ * \brief A general hierarchical slope limiter implementation.
+ *
+ * Implements hierarchical slope limiters that limit the logical derivatives of
+ * the `Tags` starting at the highest derivatives. The simplest of these
+ * limiters is described in Krivodonova 2007 \cite Krivodonova2007. The
+ * Krivodonova limiter is applied to the spectral coefficients. In 3d the
+ * function being limited is expanded as:
+ *
+ * \f{align}{
+ * u^{l,m,n}=\sum_{i,j,k=0,0,0}^{N_i,N_j,N_k}c^{l,m,n}_{i,j,k}
+ *  P_{i}(\xi)P_{j}(\eta)P_{k}(\zeta)
+ * \f}
+ *
+ * where \f$\left\{\xi, \eta, \zeta\right\}\f$ are the logical coordinates,
+ * \f$P_{i}\f$ the basis functions (typically Legendre polynomials), and the
```

Comment:
In `Spectral` the basis functions are denoted with `\phi`

Reply:
So? They're the Legendre polynomials, which are `P`.

Reply:
And in Spectral they appear to be incorrectly denoted by `\Phi` instead of `\phi`. The capital ones are reserved for spacetime test functions, not spatial ones.

Reply:
you say "typically Legendre polynomials". If this is general, I think `\phi` should be used. If Legendre polynomials are indeed assumed here, then `P`.

Reply:
The original derivation and the lower bound by Krivodonova are for Legendre polynomials. However, the procedure is general for any 1d basis (could probably be generalized to 2d bases but :man_shrugging: ). I will change to `\phi` and reword to make it clear that Krivodonova used Legendre, but that the procedure is general.

Reply:
I'm leaving it as `P` focusing on Legendre polynomials and mentioning where the fact that Legendre polynomials are needed that we relax this requirement and how (the lower bound of the alphas)